### PR TITLE
[Hardware][Neuron] Add on-device sampling support for Neuron

### PR DESCRIPTION
**Pull request by chongmni-aws**
Add support for on-device sampling for Neuron. The added feature allows each batch line to be sampled dynamically, while each batch line can have unique sampling parameters.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/worker/neuron_model_runner.py [hunk 1]
```diff
@@ -179,11 +202,13 @@ def _prepare_decode(
...
-                                               max_len=1,
-                                               dtype=torch.long,
-                                               device=self.device)
+        input_positions = make_tensor_with_pad(
+            input_positions,
+            pad=0,
+            max_len=1,
+            dtype=torch.long,
+            device=self.device,
+        )
```
Comment (liangfu) [inline vllm/worker/neuron_model_runner.py:211]: please avoid unnecessary/unrelated code style changes (if possible), since it will be confusing if we change unrelated code style back and forth. if these code style changes is intentional, we can always refactor in a separate PR.

File: vllm/worker/neuron_model_runner.py [hunk 2]
```diff
@@ -201,20 +226,24 @@ def prepare_model_input(
...
-            (input_tokens, input_positions, input_block_ids, seq_lens,
-             multi_modal_kwargs
-             ) = self._prepare_prompt(seq_group_metadata_list)
+            (
+                input_tokens,
+                input_positions,
+                input_block_ids,
+                seq_lens,
+                multi_modal_kwargs,
+            ) = self._prepare_prompt(seq_group_metadata_list)
```
Comment (liangfu) [inline vllm/worker/neuron_model_runner.py:243]: avoid unnecessary/unrelated code style changes (if possible)

File: vllm/worker/neuron_model_runner.py [hunk 3]
```diff
@@ -225,13 +254,48 @@ def prepare_model_input(
...
+        # This avoids calling update_generation_config for every token within
+        # the same sequence.
+        if (top_k != current_generation_config.top_k
+                or top_p != current_generation_config.top_p
+                or temperature != current_generation_config.temperature):
+            current_generation_config.top_k = top_k
+            current_generation_config.top_p = top_p
+            current_generation_config.temperature = temperature
+            self.model_config.generation_config = copy.deepcopy(
+                current_generation_config)
```
Comment (liangfu) [inline vllm/worker/neuron_model_runner.py:295]: as this is part of `execute_model` function, `deepcopy` could introduce unnecessary overhead.

**Review by liangfu — CHANGES_REQUESTED** [review]
Thank you @chongmni-aws for contributing. I left a few comments.

File: vllm/worker/neuron_model_runner.py (hunk 2 above)
Comment (chongmni-aws) [reply vllm/worker/neuron_model_runner.py]: ack

File: vllm/worker/neuron_model_runner.py (hunk 1 above)
Comment (chongmni-aws) [reply vllm/worker/neuron_model_runner.py]: ack

File: vllm/worker/neuron_model_runner.py (hunk 3 above)
Comment (chongmni-aws) [reply vllm/worker/neuron_model_runner.py]: ack, remove the deepcopy as the config duplication is already done during the `self.model.model.update_generation_config(current_generation_config)`

File: vllm/worker/neuron_model_runner.py [hunk 4]
```diff
@@ -75,6 +76,15 @@ def __init__(
 
         # Lazy initialization.
         self.model: nn.Module  # initialize after load_model.
+        self.model_config.generation_config = GenerationConfig(
+            max_length=self.scheduler_config.max_model_len,
+            do_sample=True,
+            per_batch_line=True,
+            top_k=[1] * self.scheduler_config.max_num_seqs,
```
Comment (liangfu) [inline vllm/worker/neuron_model_runner.py:83]: The default for `sampling_params.top_k` is -1. Is there a way we can get default from SamplingParams instead of introducing another default here?

File: vllm/worker/neuron_model_runner.py [hunk 5]
```diff
@@ -75,6 +76,15 @@ def __init__(
 
         # Lazy initialization.
         self.model: nn.Module  # initialize after load_model.
+        self.model_config.generation_config = GenerationConfig(
```
Comment (liangfu) [inline vllm/worker/neuron_model_runner.py:79]: We call this `sampling_params` or `SamplingParams` in vLLM. It would be better if we can make it consistent.

File: vllm/worker/neuron_model_runner.py [hunk 6]
```diff
@@ -75,6 +76,15 @@ def __init__(
...
+            max_length=self.scheduler_config.max_model_len,
+            do_sample=True,
+            per_batch_line=True,
+            top_k=[1] * self.scheduler_config.max_num_seqs,
+            top_p=[1] * self.scheduler_config.max_num_seqs,
```
Comment (liangfu) [inline vllm/worker/neuron_model_runner.py:84]: The default for sampling_params.top_p is 1.0 (floating-point) instead 1.

File: vllm/worker/neuron_model_runner.py [hunk 7]
```diff
@@ -233,6 +243,36 @@ def prepare_model_input(
...
+
+        # Only call update the generation config is the new config is different.
+        # This avoids calling update_generation_config for every token within
+        # the same sequence.
+        if (top_k != current_generation_config.top_k
+                or top_p != current_generation_config.top_p
+                or temperature != current_generation_config.temperature):
+            current_generation_config.top_k = top_k
+            current_generation_config.top_p = top_p
+            current_generation_config.temperature = temperature
```
Comment (liangfu) [inline vllm/worker/neuron_model_runner.py:271]:
```
top_k=[1] * self.scheduler_config.max_num_seqs,
```
top_k is initialized as a vector. Are we trying to override with a scalar value here ?

**Review by liangfu — CHANGES_REQUESTED** [review]
Thanks @chongmni-aws for the update. The change looks good overall, and I left a few more comments concerning consistency.

File: vllm/worker/neuron_model_runner.py (hunk 7 above)
Comment (chongmni-aws) [reply vllm/worker/neuron_model_runner.py:271]: nope, here the `top_k` is also a vector as it's a copy of the original vector `top_k = current_generation_config.top_k.copy()`

File: vllm/worker/neuron_model_runner.py (hunk 4 above)
Comment (chongmni-aws) [reply vllm/worker/neuron_model_runner.py:83]: we can change it to -1, but the `top_k`, `top_p`, and `temperature` here don't really take any effect as their value will be override with the values in SamplingParams during the `execute_model` call.

File: vllm/worker/neuron_model_runner.py (hunk 6 above)
Comment (chongmni-aws) [reply vllm/worker/neuron_model_runner.py:84]: ack.

File: vllm/worker/neuron_model_runner.py (hunk 5 above)
Comment (chongmni-aws) [reply vllm/worker/neuron_model_runner.py:79]: ack, update it.

File: vllm/model_executor/model_loader/neuron.py [hunk 8]
```diff
@@ -71,8 +74,23 @@ def sample(
         logits: torch.Tensor,
         sampling_metadata: SamplingMetadata,
     ) -> Optional[SamplerOutput]:
-        next_tokens = self.sampler(logits, sampling_metadata)
```
Comment (xendo) [inline vllm/model_executor/model_loader/neuron.py:74]: you are not longer using sampler, can you remove it from the class?

File: vllm/worker/neuron_model_runner.py [hunk 9]
```diff
@@ -233,6 +247,39 @@ def prepare_model_input(
...
+        top_k = current_sampling_params.top_k.copy()
+        top_p = current_sampling_params.top_p.copy()
+        temperature = current_sampling_params.temperature.copy()
+        for index, sequence_group_to_sample in enumerate(
+                sampling_metadata.seq_groups):
+            top_k[index] = self._convert_to_neuron_top_k(
+                sequence_group_to_sample.sampling_params.top_k)
+            top_p[index] = sequence_group_to_sample.sampling_params.top_p
+            temperature[index] = \
+                sequence_group_to_sample.sampling_params.temperature
```
Comment (xendo) [inline vllm/worker/neuron_model_runner.py:266]: We are only applying 3 out of couple of dozen sampling parameters available, it's definitely not a backward compatible change. Can we at least have a switch to turn it off? Would be also good to warn which parameters were not applied.

File: vllm/worker/neuron_model_runner.py [hunk 10]
```diff
@@ -233,6 +247,39 @@ def prepare_model_input(
...
+
+        # Only update the sampling parameters if there is a change.
+        if (top_k != current_sampling_params.top_k
```
Comment (xendo) [inline vllm/worker/neuron_model_runner.py:269]: wouldn't sampling params change only when a new sequence group enters a batch? it may be more efficient way of checking than going through all params one by one

File: vllm/worker/neuron_model_runner.py [hunk 11]
```diff
@@ -245,6 +292,9 @@ def execute_model(
             raise ValueError(
                 "NeuronModelRunner does not support multi-step execution.")
 
+        # Update Neuron's generation configs from sampling_metadata
```
Comment (xendo) [inline vllm/worker/neuron_model_runner.py:295]: nit. the comment doesn't add anything beyond what the method name already says

File: vllm/model_executor/model_loader/neuron.py (hunk 8 above)
Comment (chongmni-aws) [reply vllm/model_executor/model_loader/neuron.py:74]: ack, removed.

File: vllm/worker/neuron_model_runner.py (hunk 11 above)
Comment (chongmni-aws) [reply vllm/worker/neuron_model_runner.py:295]: removed.

File: vllm/worker/neuron_model_runner.py (hunk 10 above)
Comment (liangfu) [reply vllm/worker/neuron_model_runner.py:269]: i like the idea of applying the change only when a new sequence group enters a batch.

Comment (chongmni-aws) [reply vllm/worker/neuron_model_runner.py]: ack, now we will record the previous batch's request_ids and we only update the generation config if there is a change in the request_ids.

File: vllm/worker/neuron_model_runner.py (hunk 4 above)
Comment (chongmni-aws) [reply vllm/worker/neuron_model_runner.py]: also -1 is not support for on-device sampling.

File: vllm/worker/neuron_model_runner.py (hunk 5 above)
Comment (chongmni-aws) [reply vllm/worker/neuron_model_runner.py]: just to be clear the sampling params here is for neuron on-device specifically, so rename it as `neuron_sampling_params`.

File: vllm/worker/neuron_model_runner.py (hunk 9 above)
Comment (chongmni-aws) [reply vllm/worker/neuron_model_runner.py:282]: hi @xendo , neuron will turn on the on device sampling as the default sampling approach to get higher perf. currently, the on-device sampling only supports top_k, top_p and temperature. we should provide a way for user to turn off the on-device sampling. @liangfu , do you think we should go with a environment variable like `NEURON_ON_DEVICE_SAMPLING_DISABLED`?

Comment (chongmni-aws) [reply vllm/worker/neuron_model_runner.py:296]: added a switch to turn on-device sampling off.

Comment (xendo) [reply vllm/worker/neuron_model_runner.py:296]: thanks!

File: vllm/worker/neuron_model_runner.py [hunk 12]
```diff
@@ -76,6 +81,33 @@ def __init__(
...
 
+        # Once NEURON_ON_DEVICE_SAMPLING_DISABLED is set to True, turn off
+        # on-device sampling
+        self._on_device_sampling_disabled = os.getenv(
+            "NEURON_ON_DEVICE_SAMPLING_DISABLED", False) == 'True'
+
+        # NEURON needs to update sampling parameters when request IDs change
+        # across batches. This variable stores the previous batch's request IDs
+        # to determine if an update is needed.
+        self._previous_batch_request_ids: List[str] = []
```
Comment (xendo) [inline vllm/worker/neuron_model_runner.py:92]: did you verify that order of request ids is deterministic?

Comment (chongmni-aws) [reply vllm/worker/neuron_model_runner.py:92]: yes! also if there is no new sequence group but the order is changed, we will need to update the config as well, as the on-device sampling parameters are ordered.

File: vllm/worker/neuron_model_runner.py [hunk 13]
```diff
@@ -76,6 +81,33 @@ def __init__(
         # Lazy initialization.
         self.model: nn.Module  # initialize after load_model.
 
+        # Once NEURON_ON_DEVICE_SAMPLING_DISABLED is set to True, turn off
+        # on-device sampling
+        self._on_device_sampling_disabled = os.getenv(
+            "NEURON_ON_DEVICE_SAMPLING_DISABLED", False) == 'True'
```
Comment (liangfu) [inline vllm/worker/neuron_model_runner.py:87]:
```suggestion
        self._on_device_sampling_disabled = int(os.getenv(
            "NEURON_ON_DEVICE_SAMPLING_DISABLED", "0"))
```

**Review by liangfu — APPROVED** [review]
Thanks @chongmni-aws . The update LGTM.

File: vllm/worker/neuron_model_runner.py (hunk 13 above)
Comment (chongmni-aws) [reply vllm/worker/neuron_model_runner.py]: updated. thanks!

**Comment by chongmni-aws** [comment]
rebase to main.

**Comment by xendo** [comment]
Thanks for addressing all comments. LGTM

**Review by xendo — APPROVED** [review]

File: vllm/worker/neuron_model_runner.py [hunk 14]
```diff
@@ -227,12 +259,49 @@ def prepare_model_input(
...
+
+        self.model.model.update_generation_config(current_sampling_params)
+
+    def _convert_to_neuron_top_k(self, top_k: int) -> int:
+        if top_k < 0 or top_k > self._MAX_NEURON_SAMPLING_TOP_K:
+            return self._MAX_NEURON_SAMPLING_TOP_K
+        return top_k
+
     @torch.inference_mode()
     def execute_model(
```
Comment (hbikki) [inline vllm/worker/neuron_model_runner.py:306]: in execute , can we ignore calling logits processing also when sampling is enabled ?

Comment (chongmni-aws) [reply vllm/worker/neuron_model_runner.py:307]: ack!

**Review by hbikki — APPROVED** [review]
LGTM

**Review by simon-mo — APPROVED** [review]
stamp
